### PR TITLE
add FAQ entry on how to run non-nix executables

### PR DESCRIPTION
**Pull request by tejing1**
Adds an FAQ entry, primarily intended as a link target for the error message introduced in nixos/nixpkgs#269551.

Given that usecase, I do want to ask how much commitment to link stability is there on nix.dev? Is there perhaps a method to create a stable link that redirects to the proper location, so that external links can remain stable even if the site structure is changed?

### CODE REVIEW

**Comment by roberth** [comment]
I would prefer to have a separate page for this, or even a dedicated "landing URL" that redirects to whichever resource is most appropriate. Such a URL would also have a clear intent, making it very unlikely to break.

In fact, I think such a decoupling would be great, and it would immediately unblock this PR if you could add a redirect in `/_redirects`.

What's more, we could get some statistics about such errors by making the URL unique.

Perhaps

```
# Stable URLs for help with diagnosing problems
# Refer to these URLs with query parameters that provide extra context
# e.g. /diagnostic/no-ld?nixos=1 to allow for later specialisation of the docs.
# Similarly, nixpkgs may refer to /diagnostics/foo?os=${stdenv.hostPlatform.parsed.kernel.name}&cpu=${stdenv.hostPlatform.parsed.cpu.name} (or some ready-made string of parameters)
/diagnostic/no-ld /guides/faq#...
```

Not sure what's the exact syntax for the `#` fragment there.

cc @fricklerhandwerk

**Comment by roberth** [comment]
NB this is currently blocking https://github.com/NixOS/nixpkgs/pull/269551, and it would be unconstitutional to punish authors of new features for writing docs.

File: source/guides/faq.md [hunk 1]
```diff
@@ -49,6 +49,44 @@ See <https://github.com/nix-community/home-manager>
...
+  There are multiple approaches:
+  - Build from source.
+    For an introduction to this, see [](../tutorials/packaging-existing-software).
+  - Set the program's ELF metadata to include paths to libraries using [`autoPatchelfHook`](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook).
+    Do this if building from source isn't feasible.
+  - Wrap the program to run in a normal-seeming container using [`buildFHSEnv`](https://nixos.org/manual/nixpkgs/stable/#sec-fhs-environments).
+    This is a last resort, but sometimes necessary.
+    If the program downloads and runs other executables, for example.
+
+- Create a library path that only applies to unpackaged programs by using `nix-ld`.
```
Comment (infinisil) [inline source/guides/faq.md:71]:
```suggestion
- Create a library path that only applies to unpackaged programs by using [`nix-ld`](https://github.com/Mic92/nix-ld).
```

File: source/guides/faq.md [hunk 2]
```diff
@@ -49,6 +49,44 @@ See <https://github.com/nix-community/home-manager>
...
+  Note that this does not work for 32-bit executables on `x86_64` machines.
```
Comment (infinisil) [inline source/guides/faq.md:72]: This seems a bit specific, why does it not work? Is it actually due to a fundamental reason or just because of https://github.com/NixOS/nixpkgs/pull/269919?

And this would be best as a note using
```
:::{note}
Text here
:::
```

File: source/guides/faq.md [hunk 3]
```diff
@@ -49,6 +49,44 @@ See <https://github.com/nix-community/home-manager>
...
+  Add this to your `configuration.nix`, rebuild, and relogin to propagate the new environment variables:
```
Comment (infinisil) [inline source/guides/faq.md:73]:
```suggestion
  Add this to your `configuration.nix` and run `nixos-rebuild switch`:
```

File: source/guides/faq.md [hunk 4]
```diff
@@ -49,6 +49,44 @@ See <https://github.com/nix-community/home-manager>
...
+
+  ```nix
+  programs.nix-ld.enable = true;
+  programs.nix-ld.libraries = with pkgs; [
+    # Add any missing dynamic libraries for unpackaged programs
+    # here, NOT in environment.systemPackages
+  ];
+  ```
```
Comment (infinisil) [inline source/guides/faq.md:81]:
~~~suggestion
  ```
  
  Then log out and back in again.
~~~

A bit harder to miss this way.

Also why is this required at all? Ideally this wouldn't be necessary.

File: source/guides/faq.md [hunk 5]
```diff
@@ -49,6 +49,44 @@ See <https://github.com/nix-community/home-manager>
...
+
+- Run your program in the normal-seeming container made for the steam package.
+  Install `steam-run` and run your program with:
+
+  ```shell-session
+  $ steam-run <command>
+  ```
```
[the remark this answers is not in the document]
Comment (infinisil) [reply source/guides/faq.md:88]: Also we should probably have a package called something like `fhs-run` (or even a `pkgs.fhsRun.withPackages`) that doesn't depend on steam.

**Review by infinisil — CHANGES_REQUESTED** [review]
Reviewed in the docs team meeting today :)

File: source/guides/faq.md (hunk 2 above)
Comment (tejing1) [reply source/guides/faq.md:72]: I only know the `nix-ld` module doesn't try to install it for 32 bit, and the `pkgs.pkgsi686Linux.nix-ld` doesn't build. I didn't look into how difficult it would be to fix the issue.

File: source/guides/faq.md [hunk 6]
```diff
@@ -49,6 +49,44 @@ See <https://github.com/nix-community/home-manager>
 
 ## NixOS
 
+### How to run non-nix executables?
+
+NixOS cannot run dynamically linked executables intended for generic linux environments out of the box.
+This is because, by design, it does not have a global library path.
+Solutions include:
```
[the remark this answers is not in the document]
Comment (tejing1) [reply source/guides/faq.md:56]: This is further from the core issue, in my view. We could support this without being at all FHS-like. `nix-ld` is proof of that.

The core issue is that, unlike distros that shove all their libraries into a shared lookup path, we don't automatically inherit a serviceable library path from the other programs on the system, and would have to explicitly create one solely for non-nix executables, as `nix-ld` does.

Also, I think far more readers will immediately understand that this restriction is somewhat fundamental to the software architecture if we say there is no global library path. Talking about the FHS makes it sound like we just want to be different for not good reason, and don't care about the system being usable.

The way nix packages programs to find their libraries and other dependencies in a separate lookup path for each program should perhaps get an article in "concepts", which could be linked here.

File: source/guides/faq.md [hunk 7]
```diff
@@ -49,6 +49,44 @@ See <https://github.com/nix-community/home-manager>
...
+- Use the version packaged in nixpkgs, if there is one.
+  You can search available packages at <https://search.nixos.org/packages>.
+
+- Write a nix expression for the program to package it in your own configuration.
+  There are multiple approaches:
+  - Build from source.
+    For an introduction to this, see [](../tutorials/packaging-existing-software).
+  - Set the program's ELF metadata to include paths to libraries using [`autoPatchelfHook`](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook).
+    Do this if building from source isn't feasible.
+  - Wrap the program to run in a normal-seeming container using [`buildFHSEnv`](https://nixos.org/manual/nixpkgs/stable/#sec-fhs-environments).
```
[the remark this answers is not in the document]
Comment (tejing1) [reply source/guides/faq.md:67]: I was trying to avoid the word "FHS" in this message. It's jargon that's basically only known in the nixos community, because we need word for what we aren't. The rest of the world doesn't need a word for it. And this message is likely to be many reader's introduction to the fact that nixos is different in this way.

File: source/guides/faq.md (hunk 4 above)
Comment (tejing1) [reply source/guides/faq.md:86]: Reworked that so the directions for after editing `configuration.nix` are after the code block.

**Comment by tejing1** [comment]
I didn't know about `_redirects`, but unfortunately it doesn't work in `devmode`, so I can't easily test whether redirection to a specific url fragment (`#foo`) in a page is possible. I suspect it isn't, though. If I am correct in that suspicion, then having a stable url through redirects would only be feasible if it's a dedicated page.

**Comment by roberth** [comment]
`nix shell nixpkgs#netlify-cli -c netlify dev` may help, providing a local dev server that supports it.

**Comment by infinisil** [comment]
I also tried out netlify-cli earlier. It works, but it's a bit big on dependencies, and you need to also add `deno` so it doesn't give an ugly warning. But it does work, even opens the page in a browser by default:

```
$ nix-build -A build
$ nix-shell -p netlify-cli deno --run 'netlify dev -d result'
```

And I can confirm that redirects to URLs with a `#` work just fine :)

**Comment by tejing1** [comment]
Ok, then. What should the stable path be? `/landing/stub-ld`?

**Comment by tejing1** [comment]
`/diagnostic` seems odd to me, since it's not really about the main purpose (as I see it), of making a stable link. Maybe just `/permalink/stub-ld`? Can't get much clearer about what it's for than that. It's a permalink for `stub-ld` to link to.

**Review by infinisil — APPROVED** [review]

**Comment by fricklerhandwerk** [comment]
I'm really not convinced of pre-emptively creating permalinks. Every link should be a permalink: If someone decides to move a file, they should add a redirect. Eventually we should enforce this in CI and then automate creating the redirect.

**Comment by infinisil** [comment]
In this case we depend on an _anchor_ though, which for nix.dev are automatically generated from the section heading. Ideally we could have persistent anchors too, but for now we don't, so I think the /permalink workaround is fine for the time being.

**Comment by tejing1** [comment]
It's technically impossible to redirect an anchor, except by javascript shenanigans, since it isn't even sent to the server. So, if we wanted to separate this from the faq later, redirects couldn't do it without this kind of preparation.

File: _redirects [hunk 8]
```diff
@@ -25,3 +25,4 @@
 /recipes/direnv /guides/recipes/direnv 301
 /tutorials/learning-journey/sharing-dependencies /guides/recipes/sharing-dependencies 301
 /tutorials/learning-journey/packaging-existing-software /tutorials/packaging-existing-software 301
+/permalink/stub-ld /guides/faq#how-to-run-non-nix-executables 301
```
Comment (roberth) [inline _redirects:28]:
```suggestion
# incoming permalinks
/in/stub-ld /guides/faq#how-to-run-non-nix-executables 301
```

Brevity worth considering.

**Review by roberth — APPROVED** [review]

**Comment by tejing1** [comment]
Thought I should avoid the idiom "bridge this gap", on further thought. It may not be readily understandable to non-native speakers.

File: _redirects (hunk 8 above)
Comment (tejing1) [reply _redirects:28]: I feel like that's a bit too much brevity. Having it be obvious what it's for just looking at the url is more useful than it being really short, I think.

**Comment by fricklerhandwerk** [comment]
Agreed on permalinks being more accessible than JavaScipt shenanigans. I'd like to give it a last pass to ensure clarity for outside observers and then merge, if that's alright for you @tejing1. It may be a few days until I get to it though.

**Comment by infinisil** [comment]
@fricklerhandwerk Unless there's incorrect information in the PR, I'd much rather prefer follow-up PRs!

**Comment by tejing1** [comment]
Sooner is better, especially since it's blocking another PR, but I don't want to rush past the appropriate level of review, either. Do what needs to be done.

File: source/guides/faq.md [hunk 9]
```diff
@@ -49,6 +49,55 @@ See <https://github.com/nix-community/home-manager>
...
 ## NixOS
 
+### How to run non-nix executables?
+
+NixOS cannot run dynamically linked executables intended for generic linux environments out of the box.
+This is because, by design, it does not have a global library path, nor does it follow the traditional linux standard for where program files are stored on the filesystem.
+
+:::{note}
+The NixOS community uses the term "FHS" for traditional linux environments, which is short for "[Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html)".
+:::
```
Comment (fricklerhandwerk) [inline source/guides/faq.md:59]:
```suggestion
This is because, by design, it does not have a global library path, nor does it follow the [Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html) (FHS).
```

There is nothing special about the Nix community using that abbreviation, and mentioning it should be enough.

File: source/guides/faq.md [hunk 10]
```diff
@@ -49,6 +49,55 @@ See <https://github.com/nix-community/home-manager>
...
+
+There are a few ways to resolve this mismatch in environment expectations:
+
+- Use the version packaged in nixpkgs, if there is one.
```
Comment (fricklerhandwerk) [inline source/guides/faq.md:63]:
```suggestion
- Use the version packaged in Nixpkgs, if there is one.
```

File: source/guides/faq.md [hunk 11]
```diff
@@ -49,6 +49,55 @@ See <https://github.com/nix-community/home-manager>
...
+  You can search available packages at <https://search.nixos.org/packages>.
+
+- Write a nix expression for the program to package it in your own configuration.
```
Comment (fricklerhandwerk) [inline source/guides/faq.md:66]:
```suggestion
- Write a Nix expression for the program to package it for your own configuration.

```

File: source/guides/faq.md [hunk 12]
```diff
@@ -49,6 +49,55 @@ See <https://github.com/nix-community/home-manager>
...
+  There are multiple approaches to this:
+  - Build from source.
+    Many open-source programs are highly flexible at compile time in terms of where their files go.
+    For an introduction to this, see [](../tutorials/packaging-existing-software).
+  - Set the program's ELF metadata to include paths to libraries using [`autoPatchelfHook`](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook).
```
Comment (fricklerhandwerk) [inline source/guides/faq.md:71]:
```suggestion
  - Set the program's [ELF header](https://en.wikipedia.org/wiki/Executable_and_Linkable_Format) to include paths to libraries using [`autoPatchelfHook`](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook).

```

File: source/guides/faq.md [hunk 13]
```diff
@@ -49,6 +49,55 @@ See <https://github.com/nix-community/home-manager>
...
+:::
+
+There are a few ways to resolve this mismatch in environment expectations:
+
+- Use the version packaged in nixpkgs, if there is one.
+  You can search available packages at <https://search.nixos.org/packages>.
+
+- Write a nix expression for the program to package it in your own configuration.
+  There are multiple approaches to this:
+  - Build from source.
```
Comment (fricklerhandwerk) [inline source/guides/faq.md:68]:
```suggestion
  - Build from source.

```

File: source/guides/faq.md [hunk 14]
```diff
@@ -49,6 +49,55 @@ See <https://github.com/nix-community/home-manager>
...
+    Many open-source programs are highly flexible at compile time in terms of where their files go.
+    For an introduction to this, see [](../tutorials/packaging-existing-software).
+  - Set the program's ELF metadata to include paths to libraries using [`autoPatchelfHook`](https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook).
+    Do this if building from source isn't feasible.
+  - Wrap the program to run in an FHS-like environment using [`buildFHSEnv`](https://nixos.org/manual/nixpkgs/stable/#sec-fhs-environments).
```
Comment (fricklerhandwerk) [inline source/guides/faq.md:73]:
```suggestion
  - Wrap the program to run in an FHS-like environment using [`buildFHSEnv`](https://nixos.org/manual/nixpkgs/stable/#sec-fhs-environments),

```

File: source/guides/faq.md [hunk 15]
```diff
@@ -49,6 +49,55 @@ See <https://github.com/nix-community/home-manager>
...
+    This is a last resort, but sometimes necessary.
+    If the program downloads and runs other executables, for example.
```
Comment (fricklerhandwerk) [inline source/guides/faq.md:75]:
```suggestion
    This is a last resort, but sometimes necessary, for example if the program downloads and runs other executables.
```

File: source/guides/faq.md [hunk 16]
```diff
@@ -49,6 +49,55 @@ See <https://github.com/nix-community/home-manager>
...
+  ```
+
+  Then run `nixos-rebuild switch`, and log out and back in again to propagate the new environment variables.
+  (This is only necessary when enabling `nix-ld`; changes in included libraries take effect immediately on rebuild.)
+
+  :::{note}
+  `nix-ld` does not work for 32-bit executables on `x86_64` machines.
+  :::
+
+- Run your program in the FHS-like environment made for the steam package using [`steam-run`](https://nixos.org/manual/nixpkgs/stable/#sec-steam-run):
```
Comment (fricklerhandwerk) [inline source/guides/faq.md:95]:
```suggestion
- Run your program in the FHS-like environment made for the Steam package using [`steam-run`](https://nixos.org/manual/nixpkgs/stable/#sec-steam-run):
```

**Review by fricklerhandwerk — APPROVED** [review]
Only formatting for better readability of list items.

File: source/guides/faq.md (hunk 11 above)
Comment (tejing1) [reply source/guides/faq.md]: I actually think "in" is better. Users unfamiliar with the nix language often don't realize that they can locate such an expression in or alongside their `configuration.nix`, and assume they need to have a local checkout of nixpkgs in order to have their own packages.

File: source/guides/faq.md (hunk 14 above)
Comment (tejing1) [reply source/guides/faq.md]: Why make this one end in a comma, when all the other options begin with a sentence saying what the approach is, and then separate sentences with commentary?

File: source/guides/faq.md (hunk 11 above)
Comment (fricklerhandwerk) [reply source/guides/faq.md]: Makes sense, but the intent is certainly not evident from just that word. Fixed in a force push.

File: source/guides/faq.md (hunk 9 above)
Comment (tejing1) [reply source/guides/faq.md]: We're the only ones who commonly use the term at all, though, abbreviated or not. So it almost certainly needs to be introduced as new jargon to those reading this. I'm not really comfortable using the word here without a prominent introduction to what it means.

I'm a bit on the fence about whether it's prominent enough without the note. It's certainly shorter and more straightforward, which is nice, but not everyone reads linearly, and I want the introduction of the term to catch the eye.

File: _redirects (hunk 8 above)
Comment (tejing1) [reply _redirects:28]: I think `permalink` is a little better, but I don't care much either way.

`topic` *might* conflict with a future desired organizational structure. Low chance, though.

File: source/guides/faq.md (hunk 9 above)
Comment (tejing1) [reply source/guides/faq.md]: Yeah, alright, I'm good with it. The different coloring of the link is enough, I think.

File: _redirects (hunk 8 above)
Comment (infinisil) [reply _redirects:28]: I'm also for `permalink`, it expresses exactly what it's for.
